packages/cli/src/utils/project.ts

```diff
@@ -0,0 +1,42 @@
+import { Project } from '@contember/cli-common'
+import { Schema } from '@contember/schema'
+import * as esbuild from 'esbuild'
+import { join, dirname } from 'path'
+import { NodeVM } from 'vm2'
+import * as fs from 'fs/promises'
+
+export const loadSchema = async (project: Project): Promise<Schema> => {
+	const tempFile = join(process.cwd(), process.env.CONTEMBER_TEMP_SCHEMA_FILE ?? '_contember_schema.js')
+	try {
+		await fs.mkdir(dirname(tempFile), { recursive: true })
+		await esbuild.build({
+			entryPoints: [join(project.apiDir, 'index.ts')],
+			platform: 'node',
+			bundle: true,
+			treeShaking: true,
+			outfile: tempFile,
+			external: ['@contember/admin', 'react', 'react/jsx-runtime'],
```

Comment:
I don't understand why these needs to be external.

Reply:
in contember plugins, which contains both schema and admin, esbuild will try to bundle it

Reply:
I don't get it. Why would the schema definition import admin?

Reply:
because the plugin has an entrypoint, which also includes component. (and modules with components does not look like they are pure, because of Component)

Reply:
then it should either not have common entrypoint or it need to add `sideEffects: false` to package.json so that it can be treeshaked

Reply:
does not work, esbuild fails fails with
```✘ [ERROR] Could not resolve "@contember/admin"

    node_modules/@mangoweb/contember-plugins/dist/editButton/admin/EditButton.js:2:43:
      2 │ import { AnchorButton, PageLinkById } from '@contember/admin';
        ╵                                            ~~~~~~~~~~~~~~~~~~

  The path "." is not currently exported by package "@contember/admin":

    node_modules/@contember/admin/package.json:8:13:
      8 │   "exports": {
        ╵              ^
..........
```

will merge with `external`, feel free to fix it later

Reply:
If it cannot be treeshaken that the entrypoint must be fixed. You cannot randomly exclude some packages and hope that stuff will keep working.